File: 04-estructuras-datos/09-arboles-binarios-busqueda/scala/ejemplos/BstBasico.scala
enum Bst:
  case Vacio
  case Nodo(valor: Int, izquierdo: Bst = Vacio, derecho: Bst = Vacio)

object BstBasico:
  def insertar(arbol: Bst, valor: Int): Bst = arbol match
    case Bst.Vacio => Bst.Nodo(valor)
    case Bst.Nodo(v, izq, der) if valor < v => Bst.Nodo(v, insertar(izq, valor), der)
    case Bst.Nodo(v, izq, der) if valor > v => Bst.Nodo(v, izq, insertar(der, valor))
    case _ => arbol

  def main(args: Array[String]): Unit =
    // La propiedad de orden decide en qué subárbol insertar.
    val arbol = List(50, 30, 70).foldLeft[Bst](Bst.Vacio)(insertar)
    println(arbol)

File: 04-estructuras-datos/09-arboles-binarios-busqueda/java/ejemplos/BstBasico.java
public class BstBasico {
    static class Nodo {
        int valor;
        Nodo izquierdo;
        Nodo derecho;
        Nodo(int valor) { this.valor = valor; }
    }

    static Nodo insertar(Nodo nodo, int valor) {
        if (nodo == null) return new Nodo(valor);

        // Un BST coloca valores menores a la izquierda y mayores a la derecha.
        if (valor < nodo.valor) nodo.izquierdo = insertar(nodo.izquierdo, valor);
        else if (valor > nodo.valor) nodo.derecho = insertar(nodo.derecho, valor);
        return nodo;
    }

    public static void main(String[] args) {
        Nodo raiz = null;
        for (int valor : new int[]{50, 30, 70, 20, 40}) raiz = insertar(raiz, valor);
        System.out.println("Raíz: " + raiz.valor);
    }
}
